system/jlib/jfile.cpp

```diff
@@ -760,6 +760,69 @@ bool CFile::remove()
 #endif
 }
 
+static std::atomic<unsigned> renameRetries{(unsigned)-1}; // initially uninitialized, set on 1st use (in rename)
+
+void setRenameRetries(unsigned numRetries)
+{
+    renameRetries = numRetries;
+}
+
+static void doRename(const char *src, const char *dst, const char *errMsg)
+{
+    unsigned retriesLeft = renameRetries;
+    unsigned retry = 0;
+    while (true)
+    {
+        if ((unsigned)-1 == retriesLeft)
+        {
+            // NB: potentially could be >1 thread here, but that's ok.
+            try
+            {
+                retriesLeft = getGlobalConfigSP()->getPropInt("misc/@numRenameRetries");
+            }
+            catch (IException *e) // handle cases where config. not available
+            {
+                EXCLOG(e, "doRename");
+                e->Release();
+                retriesLeft = 0;
+            }
+            dbgassertex((unsigned)-1 != retriesLeft);
+            setRenameRetries(retriesLeft);
+        }
+        if (-1 != ::rename(src, dst))
+            return; // success
+
+        if (0 == retriesLeft)
+            throw makeErrnoExceptionV("%s(%s, %s)", errMsg, src, dst);
+
+#ifndef _WIN32 // NB: this is primary here to help track down the issue report in HPCC-28454
+        struct stat srcInfo;
+        if (0 != stat(src, &srcInfo))
+            throw makeErrnoExceptionV("%s-doRename - stat SRC(%s)", errMsg, src);
+        CDateTime modTime;
+        timetToIDateTime(&modTime, srcInfo.st_mtime);
+        StringBuffer srcModTime;
+        modTime.getString(srcModTime);
+
+        StringBuffer dstPath;
+        splitDirTail(dst, dstPath);
+        struct stat dstInfo;
+        if (0 != stat(dstPath, &dstInfo))
+            throw makeErrnoExceptionV("%s-doRename - stat DST-PATH(%s)", errMsg, dstPath.str());
+        timetToIDateTime(&modTime,  dstInfo.st_mtime);
+        StringBuffer dstModTime;
+        modTime.getString(dstModTime);
+        DBGLOG("%s-doRename(%s, %s) FAILED (attempt %u)", errMsg, src, dst, retry+1);
+        DBGLOG("       src stat[uid=%u, gid=%u, size=%" I64F "u, dir=%s, modtime=%s]", (unsigned)srcInfo.st_uid, (unsigned)srcInfo.st_gid, (offset_t)srcInfo.st_size, boolToStr(S_ISDIR(srcInfo.st_mode)), srcModTime.str());
+        DBGLOG("  dst-path stat[uid=%u, gid=%u, size=%" I64F "u, dir=%s, modtime=%s]", (unsigned)dstInfo.st_uid, (unsigned)dstInfo.st_gid, (offset_t)dstInfo.st_size, boolToStr(S_ISDIR(dstInfo.st_mode)), dstModTime.str());
+#endif
+        MilliSleep(100 * 2^retry); // start at 100ms and double each retry
```

Comment:
This will xor with retry!  I think what you mean is

`100 << retry`

Reply:
whoops, yes, will change.